### PR TITLE
refactor: refactor mfa tests

### PR DESCRIPTION
**Pull request by J0**
## What kind of change does this PR introduce?

The mfa tests are hard to read. There's also a lot of redundant code which makes testing for hooks quite a bit harder. This PR aims to remove some of the redundancy so that it's easier to write the tests for #1314

Main changes include
- Splitting out `enrollAndVerify` into `enroll` and `verify` 
-  Using suite specific constants
-  Packaging duplicated setup code

### CODE REVIEW

File: internal/api/mfa_test.go [hunk 1]
```diff
@@ -119,20 +128,8 @@ func (ts *MFATestSuite) TestEnrollFactor() {
 	}
 	for _, c := range cases {
 		ts.Run(c.desc, func() {
-			var buffer bytes.Buffer
-			require.NoError(ts.T(), json.NewEncoder(&buffer).Encode(map[string]string{"friendly_name": c.friendlyName, "factor_type": c.factorType, "issuer": c.issuer}))
```
Comment (J0) [inline internal/api/mfa_test.go:123]: Moved this up as it was duplicated across the setup

File: internal/api/mfa_test.go [hunk 2]
```diff
@@ -176,6 +173,8 @@ func (ts *MFATestSuite) TestChallengeFactor() {
 }
 
 func (ts *MFATestSuite) TestMFAVerifyFactor() {
+	user, err := models.FindUserByEmailAndAudience(ts.API.db, ts.TestEmail, ts.Config.JWT.Aud)
```
Comment (J0) [inline internal/api/mfa_test.go:176]: ditto

File: internal/api/mfa_test.go [hunk 3]
```diff
@@ -271,7 +269,18 @@ func (ts *MFATestSuite) TestMFAVerifyFactor() {
...
+	require.NoError(ts.T(), err)
+	session, err := models.FindSessionByUserID(ts.API.db, user.ID)
+	require.NoError(ts.T(), err)
+	return user, session
+}
+
 func (ts *MFATestSuite) TestUnenrollVerifiedFactor() {
+
+	user, session := ts.setupUserAndSession()
```
Comment (hf) [inline internal/api/mfa_test.go:282]: Maybe you can use the test suite setup pattern here? OK as it is as well.

**Review by hf — APPROVED** [review]

File: internal/api/mfa_test.go [hunk 4]
```diff
@@ -511,13 +503,35 @@ func enrollAndVerify(ts *MFATestSuite, user *models.User, token string) (verifyR
...
+
+func enrollAndVerify(ts *MFATestSuite, user *models.User, token string) *httptest.ResponseRecorder {
+	w := enroll(ts, token, "", models.TOTP, ts.TestDomain, http.StatusOK)
+	enrollResp := EnrollFactorResponse{}
+	require.NoError(ts.T(), json.NewDecoder(w.Body).Decode(&enrollResp))
+	factorID := enrollResp.ID
+
+	// Challenge
+	var challengeBuffer bytes.Buffer
+	x := httptest.NewRecorder()
```
Comment (kangmingtay) [inline internal/api/mfa_test.go:523]: i would typically name this as `w` to indicate where the response is written out to
```suggestion
	w := httptest.NewRecorder()
```

File: internal/api/mfa_test.go [hunk 5]
```diff
@@ -455,43 +460,30 @@ func signUp(ts *MFATestSuite, email, password string) (signUpResp AccessTokenRes
...
-	verifyResp = enrollAndVerify(ts, signUpResp.User, signUpResp.Token)
+	resp := enrollAndVerify(ts, signUpResp.User, signUpResp.Token)
 
-	return verifyResp
+	return resp
 
 }
 
-func enrollAndVerify(ts *MFATestSuite, user *models.User, token string) (verifyResp *AccessTokenResponse) {
+func enroll(ts *MFATestSuite, token, friendlyName, factorType, issuer string, expectedCode int) *httptest.ResponseRecorder {
```
Comment (kangmingtay) [inline internal/api/mfa_test.go:472]: maybe `performEnrollFlow` or `performEnrollRequest` might be a better name

File: internal/api/mfa_test.go [hunk 6]
```diff
@@ -455,43 +460,30 @@ func signUp(ts *MFATestSuite, email, password string) (signUpResp AccessTokenRes
...
-	req.Header.Set("Content-Type", "application/json")
-	ts.API.handler.ServeHTTP(x, req)
-	require.Equal(ts.T(), http.StatusOK, x.Code)
-	challengeResp := EnrollFactorResponse{}
-	require.NoError(ts.T(), json.NewDecoder(x.Body).Decode(&challengeResp))
-	challengeID := challengeResp.ID
-
-	// Verify
+}
+func verify(ts *MFATestSuite, challengeID, factorID uuid.UUID, token string, expectedCode int) *httptest.ResponseRecorder {
```
Comment (kangmingtay) [inline internal/api/mfa_test.go:486]: maybe `performVerifyFlow` or `performVerifyRequest` might be a better name

File: internal/api/mfa_test.go [hunk 7]
```diff
@@ -455,43 +460,30 @@ func signUp(ts *MFATestSuite, email, password string) (signUpResp AccessTokenRes
 	return data
 }
 
-func signUpAndVerify(ts *MFATestSuite, email, password string) (verifyResp *AccessTokenResponse) {
+func signUpAndVerify(ts *MFATestSuite, email, password string) *httptest.ResponseRecorder {
```
Comment (kangmingtay) [inline internal/api/mfa_test.go:463]: `performTestSignupAndVerify`?

**Review by kangmingtay — APPROVED** [review]

**Pull request merged**